hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,276 @@
+        }
+    }
+
+/*!
+ * \param timestep Current time step of the simulation
+ */
+void mpcd::ReverseNonequilibriumShearFlow::update(uint64_t timestep)
+    {
+    // if slabs have changed, update them
+    if (m_update_slabs)
+        {
+        setSlabs();
+        m_update_slabs = false;
+        }
+
+    findSwapParticles();
+    swapPairMomentum();
+    }
+
+/*!
+ * Finds all particles in the "min" and "max" slab in y direction and
+ * puts them into two GPUArrays, sorted by their momentum closest to p_target in x direction.
+ */
+void mpcd::ReverseNonequilibriumShearFlow::findSwapParticles()
+    {
+    // get needed array handles. All in read mode, no particle is changed in this search function
+    ArrayHandle<Scalar4> h_vel(m_mpcd_pdata->getVelocities(),
+                               access_location::host,
+                               access_mode::read);
+    ArrayHandle<Scalar4> h_pos(m_mpcd_pdata->getPositions(),
+                               access_location::host,
+                               access_mode::read);
+
+    // sort pairs according to their sign and momentum
+    // temporary pairs to hold particles in slab
+    // first index is momentum in x dir - second index is particle tag
+    detail::ReverseNonequilibriumShearFlowSorter cmp_hi(-m_p_target);
+    detail::ReverseNonequilibriumShearFlowSorter cmp_lo(m_p_target);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_lo_slab(cmp_lo);
+    std::map<Scalar, unsigned int, detail::ReverseNonequilibriumShearFlowSorter>
+        particles_in_hi_slab(cmp_hi);
+
+    // sort particles into their slab in y-direction and record momentum in x-direction and index
+    // injecting into std::map will sort them from lowest to highest momentum
+    for (unsigned int j = 0; j < m_mpcd_pdata->getN(); j++)
+        {
+        const Scalar4 vel = h_vel.data[j];
+        const Scalar momentum = vel.x * m_mpcd_pdata->getMass();
+        const Scalar y = h_pos.data[j].y;
+        if (m_lo_pos.x <= y && y < m_lo_pos.y
+            && momentum > Scalar(0)) // lower slab, search for max momentum
+            {
+            particles_in_lo_slab.insert(std::make_pair(momentum, j));
+            }
+        else if (m_hi_pos.x <= y && y < m_hi_pos.y
+                 && momentum < Scalar(0)) // higher slab, search for min momentum
+            {
+            particles_in_hi_slab.insert(std::make_pair(momentum, j));
+            }
+        }
+
+    // find the fastest particles (in +x) in the lo slab
+    ArrayHandle<Scalar2> h_layer_lo(m_layer_lo, access_location::host, access_mode::overwrite);
+    m_num_lo = 0;
+    for (auto iter = particles_in_lo_slab.begin(); iter != particles_in_lo_slab.end(); ++iter)
+        {
+        if (m_num_lo < m_num_swap)
+            {
+            // (x,y) = (index, momentum)
+            h_layer_lo.data[m_num_lo++] = make_scalar2(__int_as_scalar(iter->second), iter->first);
+            }
+        }
```

Comment:
I think this will allow the loop to terminate early. Does this work? If yes, do the same below.
```suggestion
    for (auto iter = particles_in_lo_slab.begin(); iter != particles_in_lo_slab.end() && m_num_lo < m_num_swap; ++iter)
        {
        // (x,y) = (index, momentum)
        h_layer_lo.data[m_num_lo++] = make_scalar2(__int_as_scalar(iter->second), iter->first);
        }
```

Reply:
Yes it works; but there wasn't any noticeable difference in computation time, at least for the chosen parameter values I worked with.

Reply:
OK, please still implement it then! At minimum, it is slightly easier to understand.

Reply:
Yeah, I did implement that